From c35b956d34bbb8bb208e49e45de2c103ca11911c Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Wed, 15 May 2013 22:38:51 +0200
Subject: [PATCH] units: rework systemd-random-seed-{load,save}.service to be a
 single service

That way ordering it with MountsRequiredFor= works properly, as this no
longer results in mount units start requests to be added to the shutdown
transaction that conflict with stop requests for the same unit.
---
 Makefile-man.am                           | 11 ++---
 Makefile.am                               | 10 ++--
 man/systemd-random-seed-load.service.xml  | 80 -------------------------------
 man/systemd-random-seed.service.xml       | 75 +++++++++++++++++++++++++++++
 units/systemd-random-seed-load.service.in | 18 -------
 units/systemd-random-seed-save.service.in | 18 -------
 units/systemd-random-seed.service.in      | 21 ++++++++
 units/systemd-tmpfiles-setup.service.in   |  1 +
 9 files changed, 104 insertions(+), 133 deletions(-)
 delete mode 100644 man/systemd-random-seed-load.service.xml
 create mode 100644 man/systemd-random-seed.service.xml
 delete mode 100644 units/systemd-random-seed-load.service.in
 delete mode 100644 units/systemd-random-seed-save.service.in
 create mode 100644 units/systemd-random-seed.service.in

diff --git a/Makefile-man.am b/Makefile-man.am
index 481423a..7d62094 100644
--- a/Makefile-man.am
+++ b/Makefile-man.am
@@ -667,16 +667,11 @@ endif
 
 if ENABLE_RANDOMSEED
 MANPAGES += \
-	man/systemd-random-seed-load.service.8
+	man/systemd-random-seed.service.8
 MANPAGES_ALIAS += \
-	man/systemd-random-seed-save.service.8 \
 	man/systemd-random-seed.8
-man/systemd-random-seed-save.service.8: man/systemd-random-seed-load.service.8
-man/systemd-random-seed.8: man/systemd-random-seed-load.service.8
-man/systemd-random-seed-save.service.html: man/systemd-random-seed-load.service.html
-	$(html-alias)
-
-man/systemd-random-seed.html: man/systemd-random-seed-load.service.html
+man/systemd-random-seed.8: man/systemd-random-seed.service.8
+man/systemd-random-seed.html: man/systemd-random-seed.service.html
 	$(html-alias)
 
 endif
diff --git a/Makefile.am b/Makefile.am
index fa626c5..8d8139c 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -3162,8 +3162,7 @@ rootlibexec_PROGRAMS += \
 	systemd-random-seed
 
 nodist_systemunit_DATA += \
-	units/systemd-random-seed-save.service \
-	units/systemd-random-seed-load.service
+	units/systemd-random-seed.service
 
 systemd_random_seed_SOURCES = \
 	src/random-seed/random-seed.c
@@ -3172,16 +3171,13 @@ systemd_random_seed_LDADD = \
 	libsystemd-label.la \
 	libsystemd-shared.la
 
-SHUTDOWN_TARGET_WANTS += \
-	systemd-random-seed-save.service
 SYSINIT_TARGET_WANTS += \
-	systemd-random-seed-load.service
+	systemd-random-seed.service
 
 endif
 
 EXTRA_DIST += \
-	units/systemd-random-seed-save.service.in \
-	units/systemd-random-seed-load.service.in
+	units/systemd-random-seed.service.in
 
 # ------------------------------------------------------------------------------
 if HAVE_LIBCRYPTSETUP
diff --git a/man/systemd-random-seed-load.service.xml b/man/systemd-random-seed-load.service.xml
deleted file mode 100644
index 693c008..0000000
--- a/man/systemd-random-seed-load.service.xml
+++ /dev/null
@@ -1,80 +0,0 @@
-<?xml version="1.0"?>
-<!--*-nxml-*-->
-<!DOCTYPE refentry PUBLIC "-//OASIS//DTD DocBook XML V4.2//EN" "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
-<!--
-  This file is part of systemd.
-
-  Copyright 2012 Lennart Poettering
-
-  systemd is free software; you can redistribute it and/or modify it
-  under the terms of the GNU Lesser General Public License as published by
-  the Free Software Foundation; either version 2.1 of the License, or
-  (at your option) any later version.
-
-  systemd is distributed in the hope that it will be useful, but
-  WITHOUT ANY WARRANTY; without even the implied warranty of
-  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
-  Lesser General Public License for more details.
-
-  You should have received a copy of the GNU Lesser General Public License
-  along with systemd; If not, see <http://www.gnu.org/licenses/>.
--->
-<refentry id="systemd-random-seed-load.service" conditional='ENABLE_RANDOMSEED'>
-
-        <refentryinfo>
-                <title>systemd-random-seed-load.service</title>
-                <productname>systemd</productname>
-
-                <authorgroup>
-                        <author>
-                                <contrib>Developer</contrib>
-                                <firstname>Lennart</firstname>
-                                <surname>Poettering</surname>
-                                <email>lennart@poettering.net</email>
-                        </author>
-                </authorgroup>
-        </refentryinfo>
-
-        <refmeta>
-                <refentrytitle>systemd-random-seed-load.service</refentrytitle>
-                <manvolnum>8</manvolnum>
-        </refmeta>
-
-        <refnamediv>
-                <refname>systemd-random-seed-load.service</refname>
-                <refname>systemd-random-seed-save.service</refname>
-                <refname>systemd-random-seed</refname>
-                <refpurpose>Load and save the system random seed at boot and shutdown</refpurpose>
-        </refnamediv>
-
-        <refsynopsisdiv>
-                <para><filename>systemd-random-seed-load.service</filename></para>
-                <para><filename>systemd-random-seed-save.service</filename></para>
-                <para><filename>/usr/lib/systemd/systemd-random-seed</filename></para>
-        </refsynopsisdiv>
-
-        <refsect1>
-                <title>Description</title>
-
-                <para><filename>systemd-random-seed-load.service</filename>
-                is an early-boot service that restores the random seed
-                of the
-                system. <filename>systemd-random-seed-save.service</filename>
-                is a late-shutdown service that saves the random seed
-                of the system. See
-                <citerefentry><refentrytitle>random</refentrytitle><manvolnum>4</manvolnum></citerefentry>
-                for details. Saving/restoring the random seed across
-                boots increases the amount of available entropy early
-                at boot. On disk the random seed is stored in
-                <filename>/var/lib/random-seed</filename>.</para>
-        </refsect1>
-
-        <refsect1>
-                <title>See Also</title>
-                <para>
-                        <citerefentry><refentrytitle>systemd</refentrytitle><manvolnum>1</manvolnum></citerefentry>,
-                        <citerefentry><refentrytitle>random</refentrytitle><manvolnum>4</manvolnum></citerefentry>
-                </para>
-        </refsect1>
-
-</refentry>
diff --git a/man/systemd-random-seed.service.xml b/man/systemd-random-seed.service.xml
new file mode 100644
index 0000000..8cd14b7
--- /dev/null
+++ b/man/systemd-random-seed.service.xml
@@ -0,0 +1,75 @@
+<?xml version="1.0"?>
+<!--*-nxml-*-->
+<!DOCTYPE refentry PUBLIC "-//OASIS//DTD DocBook XML V4.2//EN" "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
+<!--
+  This file is part of systemd.
+
+  Copyright 2012 Lennart Poettering
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+-->
+<refentry id="systemd-random-seed.service" conditional='ENABLE_RANDOMSEED'>
+
+        <refentryinfo>
+                <title>systemd-random-seed.service</title>
+                <productname>systemd</productname>
+
+                <authorgroup>
+                        <author>
+                                <contrib>Developer</contrib>
+                                <firstname>Lennart</firstname>
+                                <surname>Poettering</surname>
+                                <email>lennart@poettering.net</email>
+                        </author>
+                </authorgroup>
+        </refentryinfo>
+
+        <refmeta>
+                <refentrytitle>systemd-random-seed.service</refentrytitle>
+                <manvolnum>8</manvolnum>
+        </refmeta>
+
+        <refnamediv>
+                <refname>systemd-random-seed.service</refname>
+                <refname>systemd-random-seed</refname>
+                <refpurpose>Load and save the system random seed at boot and shutdown</refpurpose>
+        </refnamediv>
+
+        <refsynopsisdiv>
+                <para><filename>systemd-random-seed.service</filename></para>
+                <para><filename>/usr/lib/systemd/systemd-random-seed</filename></para>
+        </refsynopsisdiv>
+
+        <refsect1>
+                <title>Description</title>
+
+                <para><filename>systemd-random-seed.service</filename>
+                is a service that restores the random seed of the
+                system at early-boot and saves it at shutdown. See
+                <citerefentry><refentrytitle>random</refentrytitle><manvolnum>4</manvolnum></citerefentry>
+                for details. Saving/restoring the random seed across
+                boots increases the amount of available entropy early
+                at boot. On disk the random seed is stored in
+                <filename>/var/lib/random-seed</filename>.</para>
+        </refsect1>
+
+        <refsect1>
+                <title>See Also</title>
+                <para>
+                        <citerefentry><refentrytitle>systemd</refentrytitle><manvolnum>1</manvolnum></citerefentry>,
+                        <citerefentry><refentrytitle>random</refentrytitle><manvolnum>4</manvolnum></citerefentry>
+                </para>
+        </refsect1>
+
+</refentry>
diff --git a/units/systemd-random-seed-load.service.in b/units/systemd-random-seed-load.service.in
deleted file mode 100644
index e9156ef..0000000
--- a/units/systemd-random-seed-load.service.in
+++ /dev/null
@@ -1,18 +0,0 @@
-#  This file is part of systemd.
-#
-#  systemd is free software; you can redistribute it and/or modify it
-#  under the terms of the GNU Lesser General Public License as published by
-#  the Free Software Foundation; either version 2.1 of the License, or
-#  (at your option) any later version.
-
-[Unit]
-Description=Load Random Seed
-Documentation=man:systemd-random-seed-load.service(8) man:random(4)
-DefaultDependencies=no
-RequiresMountsFor=@RANDOM_SEED@
-After=systemd-readahead-collect.service systemd-readahead-replay.service systemd-remount-fs.service
-Before=sysinit.target final.target
-
-[Service]
-Type=oneshot
-ExecStart=@rootlibexecdir@/systemd-random-seed load
diff --git a/units/systemd-random-seed-save.service.in b/units/systemd-random-seed-save.service.in
deleted file mode 100644
index 3444d4c..0000000
--- a/units/systemd-random-seed-save.service.in
+++ /dev/null
@@ -1,18 +0,0 @@
-#  This file is part of systemd.
-#
-#  systemd is free software; you can redistribute it and/or modify it
-#  under the terms of the GNU Lesser General Public License as published by
-#  the Free Software Foundation; either version 2.1 of the License, or
-#  (at your option) any later version.
-
-[Unit]
-Description=Save Random Seed
-Documentation=man:systemd-random-seed-load.service(8) man:random(4)
-DefaultDependencies=no
-RequiresMountsFor=@RANDOM_SEED@
-After=systemd-remount-fs.service systemd-random-seed-load.service
-Before=final.target
-
-[Service]
-Type=oneshot
-ExecStart=@rootlibexecdir@/systemd-random-seed save
diff --git a/units/systemd-random-seed.service.in b/units/systemd-random-seed.service.in
new file mode 100644
index 0000000..0c21a04
--- /dev/null
+++ b/units/systemd-random-seed.service.in
@@ -0,0 +1,21 @@
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+[Unit]
+Description=Load/Save Random Seed
+Documentation=man:systemd-random-seed-load.service(8) man:random(4)
+DefaultDependencies=no
+RequiresMountsFor=@RANDOM_SEED@
+Conflicts=shutdown.target
+After=systemd-readahead-collect.service systemd-readahead-replay.service systemd-remount-fs.service
+Before=sysinit.target shutdown.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=@rootlibexecdir@/systemd-random-seed load
+ExecStop=@rootlibexecdir@/systemd-random-seed save
diff --git a/units/systemd-tmpfiles-setup.service.in b/units/systemd-tmpfiles-setup.service.in
index 4a3441c..67c7d4a 100644
--- a/units/systemd-tmpfiles-setup.service.in
+++ b/units/systemd-tmpfiles-setup.service.in
@@ -10,6 +10,7 @@ Description=Recreate Volatile Files and Directories
 Documentation=man:tmpfiles.d(5) man:systemd-tmpfiles(8)
 DefaultDependencies=no
 Wants=local-fs.target
+Conflicts=shutdown.target
 After=systemd-readahead-collect.service systemd-readahead-replay.service local-fs.target
 Before=sysinit.target shutdown.target
 ConditionDirectoryNotEmpty=|/usr/lib/tmpfiles.d
-- 
1.8.2.3

